modules/service/src/main/scala/lucuma/odb/service/VisitService.scala

```diff
@@ -36,82 +47,179 @@ trait VisitService[F[_]] {
+
+  def recordGmosNorth(
+    input: RecordGmosVisitInput[GmosNorth]
   )(using Transaction[F], Services.ServiceAccess): F[Result[Visit.Id]]
 
-  def insertGmosSouth(
-    observationId: Observation.Id,
-    static:        GmosSouth
+  def recordGmosSouth(
+    input: RecordGmosVisitInput[GmosSouth]
   )(using Transaction[F], Services.ServiceAccess): F[Result[Visit.Id]]
 
-}
 
-object VisitService {
+object VisitService:
+
+  case class ObsDescription(
+    instrument:      Instrument,
+    calibrationRole: Option[CalibrationRole]
+  ):
+    // TODO: need to make instrument and site 1:1
+    def site: Site =
+      instrument.site.head
+
+    // Determine whether an observation is chargeable.  I'd originally planned
+    // to do this by generating the sequence and dropping atoms until the first
+    // one with a chargeable obsclass was found (if ever).  That's a more
+    // general way that wouldn't need to be updated as code changes, but this is
+    // far cheaper and is correct at least for now.
+    def isChargeable: Boolean =
+      calibrationRole match
+        case None                                     |
+             Some(CalibrationRole.Photometric)        |
+             Some(CalibrationRole.SpectroPhotometric) |
+             Some(CalibrationRole.Telluric)            => true
+        case Some(CalibrationRole.Twilight)            => false
 
   def instantiate[F[_]: Concurrent](using Services[F]): VisitService[F] =
-    new VisitService[F] {
+    new VisitService[F]:
 
       override def select(
         visitId: Visit.Id
       ): F[Option[VisitRecord]] =
         session.option(Statements.SelectVisit)(visitId)
 
-      def selectAll(
+      override def selectAll(
         observationId: Observation.Id
       ): Stream[F, VisitRecord] =
         session.stream(VisitService.Statements.SelectAllVisit)(observationId, 1024)
 
-      private def insert(
+      private def obsDescription(
+        observationId: Observation.Id
+      ): ResultT[F, ObsDescription] =
+        ResultT:
+          session.option(Statements.SelectObsDescription)(observationId).map: od =>
+            Result.fromOption(od, OdbError.InvalidObservation(observationId, s"Observation '$observationId' not found.".some).asProblem)
+
+
+      override def lookupOrInsert(
+        observationId: Observation.Id
+      )(using Transaction[F], Services.ServiceAccess): F[Result[Visit.Id]] =
+
+        def obsNight(site: Site): ResultT[F, ObservingNight] =
+          ResultT.liftF(timeService.currentObservingNight(site))
+
+        // Application transaction advisory lock on visit creation. The idea is
+        // to prevent two callers (Observe and Navigate) from doing a lookup
+        // simultaneously and coming to the conclusion to each create a new
+        // visit.
+        val lockVisitCreation: ResultT[F, Unit] =
+          ResultT.liftF:
+            session.unique(Statements.LockCreation)(Statements.VisitCreationLockId)
```

Comment:
I'll just call attention to arguably the sketchiest part of the PR above.

Reply:
I have never used an advisory lock but this seems reasonable to me, since there is no way to detect this situation via a key collision.